File: frontend/src/types/presentationSnapshot.ts
export type PresentationSnapshotSection =
  | "marketing_brief"
  | "marketing_strategy"
  | "copy_matrix"
  | "video_project"
  | "render_task"
  | "artifact"
  | "publish_task"
  | "campaigns";

export interface PresentationSnapshotCreateRequest {
  marketing_brief_id: number | null;
  marketing_strategy_id: number | null;
  copy_matrix_id: number | null;
  video_project_id: number | null;
  render_task_id: number | null;
  artifact_id: number | null;
  publish_task_id: number | null;
  campaign_ids: number[];
}

export interface PresentationSnapshot {
  id: number;
  schema_version: number;
  digest: string;
  product_id: number;
  marketing_brief_id: number | null;
  marketing_strategy_id: number | null;
  copy_matrix_id: number | null;
  video_project_id: number | null;
  render_task_id: number | null;
  artifact_id: number | null;
  publish_task_id: number | null;
  campaign_ids: number[];
  missing_sections: PresentationSnapshotSection[];
  snapshot_payload: Record<string, unknown>;
  artifact_sha256: string | null;
  artifact_snapshot_path: string | null;
  created_at: string;
}

export interface PresentationSnapshotCreateResult {
  snapshot: PresentationSnapshot;
  reused: boolean;
  provider_calls: 0;
}

File: backend/app/schemas/presentation_snapshot.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class PresentationSnapshotCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    marketing_brief_id: int | None = Field(default=None, ge=1)
    marketing_strategy_id: int | None = Field(default=None, ge=1)
    copy_matrix_id: int | None = Field(default=None, ge=1)
    video_project_id: int | None = Field(default=None, ge=1)
    render_task_id: int | None = Field(default=None, ge=1)
    artifact_id: int | None = Field(default=None, ge=1)
    publish_task_id: int | None = Field(default=None, ge=1)
    campaign_ids: list[int] = Field(default_factory=list)

    @field_validator("campaign_ids")
    @classmethod
    def validate_campaign_ids(cls, value: list[int]) -> list[int]:
        if any(campaign_id < 1 for campaign_id in value):
            raise ValueError("Campaign IDs must be positive")
        if len(value) != len(set(value)):
            raise ValueError("Campaign IDs must be unique")
        return value


class PresentationSnapshotRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: int
    schema_version: int
    digest: str
    product_id: int
    marketing_brief_id: int | None
    marketing_strategy_id: int | None
    copy_matrix_id: int | None
    video_project_id: int | None
    render_task_id: int | None
    artifact_id: int | None
    publish_task_id: int | None
    campaign_ids: list[int]
    missing_sections: list[str]
    snapshot_payload: dict[str, object]
    artifact_sha256: str | None
    artifact_snapshot_path: str | None
    created_at: datetime


class PresentationSnapshotCreateRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    snapshot: PresentationSnapshotRead
    reused: bool
    provider_calls: Literal[0] = 0
